### PR TITLE
Added stake and redeem and auditor_lock and unlock

### LINKED ISSUE #6: Support fund Staking and Redeem for Bank pallet
Bank should provide function via traits to:
Stake some funds (from free -> reserved)
Redeem these funds( from reserved -> free).
Staking is instant, Redeeming requires time, during which the funds are "locked", and the length (defined in blocks) is defined in Config.

Accounts must only be reaped if the total amount (free + reserved + locked) are below ED.

Provide an extrinsic for users to stake and redeem funds.
Users can redeem multiple times at the same time (multiple sets of funds locked, waiting to be redeemed).

Provide an extrinsic for Auditors to lock other customers' funds. Funds are taken from "free" first, then from "reserved". A length must be given as part of the call.

Provide an extrinsic for Auditor to "unlock" ONLY funds locked by an Auditor. These locks should have an ID associated. Any auditor can unlock funds locked by other Auditors.

### PR DESCRIPTION
**Pull request by IRISxiaowang**
Added stake and redeem and auditor_lock and unlock.

Closes #6

### CODE REVIEW

File: pallets/pallet-bank/src/lib.rs [hunk 1]
```diff
@@ -108,6 +132,37 @@ pub mod module {
...
+        },
+
+        /// Auditor unlock some fund from a user who was locked the fund by an auditor role.
+        AuditorUnlock {
+            user: T::AccountId,
+            amount: T::Balance,
+        },
+
+        /// Unlock some fund from an account's "locked" to "free".
+        UnLock {
```
Comment (syan095) [inline pallets/pallet-bank/src/lib.rs]: Unlock

File: pallets/pallet-bank/src/lib.rs [hunk 2]
```diff
@@ -167,6 +232,9 @@ pub mod module {
 
         fn on_finalize(_block_number: BlockNumberFor<T>) {
```
Comment (syan095) [inline pallets/pallet-bank/src/lib.rs]: _name is used for when a variable isn't used.
If it's used, the _ should be removed.

File: pallets/pallet-bank/src/lib.rs [hunk 3]
```diff
@@ -167,6 +232,9 @@ pub mod module {
...
             Self::reap_accounts();
+            AccountWithUnlockedFund::<T>::take(&_block_number).iter().for_each(|(user, lock_id)|{
```
Comment (syan095) [inline pallets/pallet-bank/src/lib.rs]: nitpick: since the result from the "take" is thrown away, I think into_iter can be used here.

File: pallets/pallet-bank/src/lib.rs [hunk 4]
```diff
@@ -267,6 +433,59 @@ impl<T: Config> BasicAccounting<T::AccountId, T::Balance> for Pallet<T> {
     }
 }
 
+impl<T: Config> Stakable<T::AccountId, T::Balance> for Pallet<T> {
+    fn stake_funds(user: &T::AccountId, amount: T::Balance) -> DispatchResult {
+        // Stake funds from free to reserved
+        T::RoleManager::ensure_role(&user, Role::Customer)?;
+        Accounts::<T>::mutate(&user, |balance| -> DispatchResult {
+            if balance.free >= amount {
```
Comment (syan095) [inline pallets/pallet-bank/src/lib.rs]: Use ensure! instead of if ... else

File: pallets/pallet-bank/src/lib.rs [hunk 5]
```diff
@@ -267,6 +433,59 @@ impl<T: Config> BasicAccounting<T::AccountId, T::Balance> for Pallet<T> {
...
+        Ok(())
+    }
+
+    fn redeem_funds(user: &T::AccountId, amount: T::Balance) -> DispatchResult {
+        // Redeem funds from reserved to free after a certain time
+        // get unlock BlockNumber
+        let unlock = T::BlockNumberProvider::current_block_number() + T::RedeemPeriod::get();
+
+        // Add new locked funds to user's Account Data
+        Accounts::<T>::mutate(&user, |balance| -> DispatchResult {
```
Comment (syan095) [inline pallets/pallet-bank/src/lib.rs]: better call this "account" instead of "balance". Balance usually represents a number, account can represent more complex type

File: pallets/pallet-bank/src/lib.rs [hunk 6]
```diff
@@ -267,6 +433,59 @@ impl<T: Config> BasicAccounting<T::AccountId, T::Balance> for Pallet<T> {
...
+            if balance.reserved >= amount {
```
Comment (syan095) [inline pallets/pallet-bank/src/lib.rs]: Again, ensure! here is better than if ... else ...

File: pallets/pallet-bank/src/lib.rs [hunk 7]
```diff
@@ -303,8 +522,9 @@ impl<T: Config> Pallet<T> {
     fn check_total_issuance() -> bool {
         TotalIssuance::<T>::get()
             == Accounts::<T>::iter()
-                .map(|(_, account)| account.free + account.reserved)
-                .sum()
+                .map(|(_, account)| account.free
```
Comment (syan095) [inline pallets/pallet-bank/src/lib.rs]: I think accountData has a total() function that can be used here

File: pallets/pallet-bank/src/lib.rs [hunk 8]
```diff
@@ -328,4 +548,33 @@ impl<T: Config> Pallet<T> {
...
+        NextLockId::<T>::mutate(|id| {
+            *id += 1u64;
+            *id
+        })
+    }
+
+    ///Transfer locked funds to free funds
+    fn unlock(account_id: &T::AccountId, locked_id: LockId) {
+        Accounts::<T>::mutate(account_id, |account_data| {
+            if let Some(index) = account_data
```
Comment (syan095) [inline pallets/pallet-bank/src/lib.rs:560]: I think part of the logic is similar to how auditor unlocks funds. Maybe you can refactor these 2 function with shared code.

File: pallets/pallet-bank/src/tests.rs [hunk 9]
```diff
@@ -219,13 +159,170 @@ fn can_reaped() {
...
+                }
+            );
+
+            assert!(Bank::check_total_issuance());
+
+            assert_noop!(
+                Bank::redeem_funds(RuntimeOrigin::signed(ALICE), 801),
+                Error::<Runtime>::InsufficientBalance
+            );
+        });
```
Comment (syan095) [inline pallets/pallet-bank/src/tests.rs:221]: There is a System::assert_last_event() function you can use

File: pallets/pallet-bank/src/tests.rs [hunk 10]
```diff
@@ -219,13 +159,170 @@ fn can_reaped() {
...
+}
+
+#[test]
+fn can_auditor_lock_funds() {
```
Comment (syan095) [inline pallets/pallet-bank/src/tests.rs]: auditor_can_lock_funds

File: pallets/pallet-bank/src/tests.rs [hunk 11]
```diff
@@ -219,13 +159,170 @@ fn can_reaped() {
...
+                    }]
+                }
+            );
+
+            assert!(Bank::check_total_issuance());
+
+            assert_noop!(
+                Bank::lock_funds_auditor(RuntimeOrigin::signed(charlie), ALICE, 801, 20),
+                Error::<Runtime>::InsufficientBalance
+            );
```
Comment (syan095) [inline pallets/pallet-bank/src/tests.rs:264]: test for incorrect role (only auditor can call this function

File: pallets/pallet-bank/src/tests.rs [hunk 12]
```diff
@@ -219,13 +159,170 @@ fn can_reaped() {
...
+        });
+}
+
+#[test]
+fn can_auditor_unlock_funds() {
```
Comment (syan095) [inline pallets/pallet-bank/src/tests.rs]: auditor_can_unlock_funds

File: pallets/pallet-bank/src/tests.rs [hunk 13]
```diff
@@ -219,13 +159,170 @@ fn can_reaped() {
...
+                    locked: vec![LockedFund {
+                        id: 1,
+                        amount: 200,
+                        reason: LockReason::Auditor,
+                    }]
+                }
+            );
+
+            System::reset_events();
+            assert_ok!(Bank::unlock_funds_auditor(RuntimeOrigin::signed(charlie),
```
Comment (syan095) [inline pallets/pallet-bank/src/tests.rs]: forgot to cargo fmt

File: pallets/pallet-bank/src/tests.rs [hunk 14]
```diff
@@ -219,13 +159,170 @@ fn can_reaped() {
...
+            assert_eq!(
+                Accounts::<Runtime>::get(&ALICE),
+                AccountData {
+                    free: 200,
+                    reserved: 800,
+                    locked: vec![]
+                }
+            );
+            assert!(Bank::check_total_issuance());
+        });
```
Comment (syan095) [inline pallets/pallet-bank/src/tests.rs:399]: test some edge cases:
* other auditor can unlock.
* Can unlock funds that doesn't exist (doesn't crash)
* Can't unlock Redeem locks.

**Review by syan095 — CHANGES_REQUESTED** [review]
Overall very good.
Some minor nitpicks and missing tests

**Review by syan095 — APPROVED** [review]
